File: solutions/Python/space-age/4/space_age.py
EARTH_YEAR_IN_SECONDS = 31557600.0
PLANET_YEARS = {
    'earth': 1.0,
    'mercury': 0.2408467,
    'venus': 0.61519726,
    'mars': 1.8808158,
    'jupiter': 11.862615,
    'saturn': 29.447498,
    'uranus': 84.016846,
    'neptune': 164.79132
}

class SpaceAge(object):
    """Age calculator for different planets."""
    def __init__(self, seconds):
        self.seconds = seconds
        for planet in PLANET_YEARS:
            setattr(self, f"on_{planet}", self._create_method(planet))
            
    def _on_planet(self, planet):
        return round(self.seconds / (EARTH_YEAR_IN_SECONDS * PLANET_YEARS[planet]), 2)
        
    def _create_method(self, planet):
        return lambda: self._on_planet(planet)
        
    def __getattr__(self, name):
        raise AttributeError(f"Planet not in SpaceAge '{name}'")

File: solutions/C/space-age/4/space_age.c
#include "space_age.h"
#include <stdio.h>
#include <stdlib.h>

#define SECONDS_IN_A_YEAR 31557600.0F 


float age(planet_t planet, int64_t seconds){

    switch (planet)
    {
        case MERCURY:
        return (seconds / SECONDS_IN_A_YEAR) / 0.2408467F;
        case VENUS:
        return (seconds / SECONDS_IN_A_YEAR) / 0.61519726F;
        case EARTH:
        return (seconds / SECONDS_IN_A_YEAR) * 1;
        case MARS:
        return (seconds / SECONDS_IN_A_YEAR) / 1.8808158F;
        case JUPITER:
        return (seconds / SECONDS_IN_A_YEAR) / 11.862615F;
        case SATURN:
        return (seconds / SECONDS_IN_A_YEAR) / 29.447498F;
        case URANUS:
        return (seconds / SECONDS_IN_A_YEAR) / 84.016846F;
        case NEPTUNE:
        return (seconds / SECONDS_IN_A_YEAR) / 164.79132F;
        default:
        return -1;
    }
} 

